### PR TITLE
fix(view): reject setElement after destruction

### PR DESCRIPTION
**Pull request by paulfalgout**
Related #131

## Linked issue

- Related #131

## What

- Make base `View#setElement` and `CollectionView#setElement` reject calls after destruction with stable diagnostic `MN0029`.
- Guard before element inspection, event delegation changes, DOM mutation, or lifecycle-state recomputation.
- Document the v5 lifecycle contract and add parameterized regression coverage for both view types.

## Why

Destroyed view instances are terminal in v5. Allowing `setElement` to recycle them could silently rebind DOM behavior and mutate lifecycle state after teardown, contradicting that contract.

## Agent-development failure addressed

An agent could previously treat a destroyed View or CollectionView as reusable by swapping its element. The stable diagnostic makes that invalid transition explicit and gives the agent a documented remediation.

## Public behavior contract

- Canonical behavior after this PR: base `View#setElement` and `CollectionView#setElement` throw `MN0029` once destruction begins, including calls during `before:destroy` and repeated, omitted, same-element, malformed, selector, and replacement-element calls after destruction.
- Superseded behavior removed: destroyed base instances no longer replace their element or redelegate View and Behavior DOM events.
- Compatibility exception and removal condition, if any: none. A custom nondelegating override owns its behavior; an override that delegates reaches the guarded base contract.

## Scope

- Impacts the main Marionette package's base View and CollectionView implementations, lifecycle documentation, diagnostic catalog, and generated distributions.
- Alive and constructor-time `setElement` behavior is unchanged.
- Region lifecycle behavior and unrelated post-destroy operations are intentionally unchanged.

## Runtime cost

- [ ] Static-only change
- [ ] Development-only runtime change
- [x] Changes an existing production path
- [ ] Adds opt-in production behavior

Measured impact:

- Bundle: cumulative measured size is 61.43 kB / 75.00 kB; exact-base deltas will be reported by CI for the amended head.
- Startup/hot path: one destroyed-state boolean branch on each existing base `setElement` call; error allocation occurs only on the invalid destroyed path.
- Allocations/retention: no new allocation or retention on valid calls.

## Deployment Impact

No website publication, npm publication, tag, release, or separate deployment is part of this PR. The behavior will ship through the normal main-package release process; no other package or application redeploy is required by this repository change alone.

## Validation

- [x] Unit tests
- [x] Source validation
- [x] Distribution validation
- [x] Package-consumer fixtures
- [x] Documentation examples, build, and links
- [x] Diagnostic catalog validation
- [x] Lint
- [x] Bundle-size budget

Commands run on exact head `063d2eaadc2cb7191bebf29cc789893917117065`:

- `npm test` — 90 files, 1,629 tests passed
- `npm run lint:ci`
- `npm run check:diagnostics` — 29 entries validated
- `npm run docs:check` — 41 pages and 384 internal links validated
- `npm run check:dist`
- `npm run test:source`
- `npm run test:fixtures`
- `npm run test:agent-benchmark-contract` — 19 tests passed
- `npm run size` — 61.43 kB / 75.00 kB
- `git diff --check marionettejs/master...HEAD`

Edge cases cover calls during `before:destroy`, repeated destruction, omitted arguments, current-element calls, malformed objects without inspection, missing selectors, replacement elements, stable class-specific diagnostics, unchanged element/DOM/state/delegation, and delegating versus nondelegating overrides.

## Package and browser impact

- Main package and generated ESM, CommonJS, UMD, and minified UMD projections change.
- No entrypoint or dependency-graph change.
- Browser consumers receive the same destroyed-state guard through all supported distributions.

## Rollback or deprecation

This is an intentional v5 lifecycle correction with no compatibility fallback. Rolling it back would restore unsafe destroyed-instance reuse.

## Review notes

The base View implementation lives in the historical `modules/view-region.js` source bundle. Direct inspection confirmed both guards are the literal first operation and destruction does not call `setElement` internally.

### CODE REVIEW

**Comment by paulfalgout** [comment]
Claude exact-head review: **APPROVE** — no blocking defects in the literal patch. Reviewed commit `8443b1b329bd71ac9dd67e805b47e215fd572d99`, including both leading guards, parameterized edge-case coverage, diagnostic contract, and exact-head validation. Non-blocking follow-ups were limited to broader lifecycle integration ideas outside this slice. Auditable artifact: `/Users/paulfalgout/.ai-reviews/20260830T164021Z-review-with-claude.md`.

**Comment by paulfalgout** [comment]
**Approved** at `8443b1b3` (the SHA I fetched). Bundle size passes: +152 B aggregate (61.38 kB against the 75.00 kB ceiling), under the 1% growth threshold, so no maintainer approval needed.

## No ordering hazard here

#302 needed the WeakMap dance because `destroy()` routed through the very methods it was guarding. I checked whether `setElement` has the same problem, and it does not: the only internal callers of `View`/`CollectionView` `setElement` are the two constructors (`view-region.js:794` and `collection-view.js:93`). The `behavior.setElement()` calls in `mixins/behaviors.js` are `Behavior`'s own method on a different class. Nothing on a teardown path reaches either guard, so a plain `_isDestroyed` check is sufficient and no authorization mechanism is warranted.

## The `tagRead` probe is the best thing in this test

```js
const tagRead = this.sinon.spy(() => { throw new Error('element inspected'); });
const malformed = Object.create(null, {
  [Symbol.toStringTag]: { get: tagRead },
});
```

Passing an element whose tag getter throws turns "did the guard run before touching the argument?" into something observable. Asserting the throw alone would pass even if `_validateEl` had already inspected the element first; this fails unless the guard is genuinely ahead of inspection, which is exactly what the description claims.

The rest of the coverage maps onto the stated rationale rather than restating the throw. The concern was that `setElement` could *silently rebind DOM behavior* on a destroyed instance — so `viewAction` and `behaviorAction` never firing proves events were not re-delegated, and `getUI('action')` still throwing MN0023 proves UI bindings were not re-established. Those two are the ones that would actually catch a regression in the behavior the PR exists to prevent.

Parameterizing over both classes, and the override pair — a non-delegating `setElement` override keeps control and returns its own value, a delegating one throws — matches the contract #302 established. Good to see the milestone converging on one story for overrides instead of a per-PR answer.

## Non-blocking: the lone `'ViewError'` literal is a trap

`view-region.js` hosts both `Region` and `View`, and its file-level `classErrorName` is `'RegionError'`. The new View guard therefore hardcodes `name: 'ViewError'` — correct, and the only occurrence of that literal in the file. Meanwhile `mixins/view.js:17` already declares `const classErrorName = 'ViewError'`, so the right value exists as a constant, just in the other half of `View`'s definition.

The hazard is a future contributor normalizing the apparent inconsistency by swapping the literal for the local `classErrorName`, which would silently relabel View errors as `RegionError`. That would be caught — the spec asserts ``name: `${name}Error` `` — so this is about the trap being avoidable rather than undetectable. A local `const viewClassErrorName = 'ViewError';` beside the View section, or importing the mixin's constant, would make the intent self-evident at zero runtime cost.

## The `/errors/` convention is now compounding

MN0029 uses `url: 'errors/MN0029/'`, making it the second diagnostic on that pattern after MN0028 — three call sites total. It is becoming the de facto style for new codes while the other 27 still point at `<page>.html#anchor`, and there is still no page source under `/errors/` and nothing in the repo consuming `docsAnchor`.

No objection to the direction; it is the better long-term shape. But each new code widens the split, and the decision is cheaper to act on now than after a dozen more. Either the generator lands and the old URLs migrate, or new codes adopt the anchor style until it does.

Nothing blocking.

**Comment by paulfalgout** [comment]
Claude exact-head review: **APPROVE** — no blocking issues on amended commit `063d2eaadc2cb7191bebf29cc789893917117065`. The review covered the symmetric `_isDestroying || _isDestroyed` guards, the parameterized `before:destroy` regression, documentation/catalog consistency, and exact-head validation. Claude noted the general uncaught lifecycle-handler exception behavior as non-blocking; Marionette's existing destroy catch clears `_isDestroying` before rethrowing. Auditable artifact: `/Users/paulfalgout/.ai-reviews/20260830T165027Z-review-with-claude.md`.

File: modules/view-region.js [hunk 1]
```diff
@@ -812,6 +812,15 @@ assignOwn(View.prototype, ViewMixin, RegionsMixin, {
   cidPrefix: 'mnv',
 
   setElement(element) {
+    if (this._isDestroyed) {
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply modules/view-region.js]: Fixed on 063d2ea: both base implementations now guard `_isDestroying || _isDestroyed` before inspecting the argument or changing delegation/state. Added a parameterized `before:destroy` regression for View and CollectionView that proves malformed input is not inspected, element identity is preserved, and destruction completes. Updated the diagnostic remediation and lifecycle/migration/changelog wording to "once destruction begins"; exact-head validation is green.

**Comment by paulfalgout** [comment]
**Approval stands** at `063d2eaa`, and this push fixes a real gap I missed. Bundle size still passes: +204 B aggregate (61.43 kB against the 75.00 kB ceiling), largest artifact delta +0.45%.

## The `_isDestroying` window was a genuine hole

I checked the teardown path for *internal* callers last time and concluded a plain `_isDestroyed` check was sufficient. That was incomplete — it ignored user handlers running inside the destroy sequence. `destroy()` sets `_isDestroying = true` and only reaches `_isDestroyed = true` after `triggerMethod('before:destroy', ...)` returns, so a `before:destroy` handler calling `setElement` passed the old guard cleanly. It would have swapped the element and redelegated events, and destruction would then have run against the *new* element while the original stayed bound and uncleaned — the exact failure the PR set out to prevent, reachable through the one window the guard did not cover.

I should have caught this: it is the same window #302's `_isDestroying` reentrancy guard exists for, so the concept was already in front of me.

The test isolates it well. Having the handler swallow the MN0029 lets `destroy()` complete normally, so the assertions are about the guard rather than about the retryable-destroy abort path, and reusing the `Symbol.toStringTag` probe proves the guard precedes inspection in this window too.

## Non-blocking: `setElement` is now the only operation guard with these semantics

Surveying the flags across the milestone:

| Guard | Flags |
| --- | --- |
| `View#destroy`, `Region#destroy`, `destroy` mixin | `_isDestroyed \|\| _isDestroying` |
| **`View#setElement`, `CollectionView#setElement`** | **`_isDestroyed \|\| _isDestroying`** |
| `View#render`, `CollectionView#render` | `_isDestroyed` only |
| `assertRegionIsLive` — MN0028 for `show`/`empty`/`reset` | `_isDestroyed` only |

Using both flags is correct for the destroy entry points, since that is reentrancy protection. But `setElement` is now the only *operation* guard that rejects from the moment destruction begins; `render` and all three Region operations still permit the `before:destroy` window.

The question that leaves for a caller — or an agent — is why replacing the element during `before:destroy` is invalid while re-rendering during `before:destroy` is fine. Both schedule work that teardown then has to undo. There may be a deliberate answer: showing a farewell View from `before:destroy` is a plausible pattern in a way that swapping the root element is not. If so, that distinction is worth stating, because it is not inferable from the guards.

I am not asking for it here — widening the Region guards would reopen the authorization design from #302, which is not a change to make in passing. But #131 is about making the destroyed-instance contract deterministic, and it currently has two different answers to "when does destruction begin" depending on which method you call.

Nothing blocking.
